docs/architecture/adr-054-semver-compatible-modules.md

```diff
@@ -0,0 +1,493 @@
+mostly or completely in the direction of inter-module communication as described in [ADR 033](./adr-033-protobuf-inter-module-comm.md).
+In this paradigm, a module could generate all the types it needs internally - including the API types of other modules -
+and talk to other modules via a client-server boundary. For instance, if `bar` needs to talk to `foo`, it could
+generate its own version of `MsgDoSomething` as `bar/internal/foo/v1.MsgDoSomething` and just pass this to the
+inter-module router which would somehow convert it to the version which foo needs (ex. `foo/internal.MsgDoSomething`).
+
+Currently, two generated structs for the same protobuf type cannot exist in the same go binary without special
+build flags (see https://developers.google.com/protocol-buffers/docs/reference/go/faq#fix-namespace-conflict).
+A relatively simple mitigation to this issue would be to set up the protobuf code to not register protobuf types
+globally if they are generated in an `internal/` package. This will require modules to register their types manually
+with the app-level level protobuf registry, this is similar to what modules already do with the `InterfaceRegistry`
+and amino codec.
+
+If modules _only_ do ADR 033 message passing then a naive and non-performant solution for converting `bar/internal/foo/v1.MsgDoSomething`
+to `foo/internal.MsgDoSomething` would be marshaling and unmarshaling in the ADR 033 router. This would break down if
+we needed to expose protobuf types in `Keeper` interfaces because the whole point is to try to keep these types
+`internal/` so that we don't end up with all the import version incompatibilities we've described above. However,
+because of the issue with minor version incompatibilities and the need for [unknown field filtering](./adr-020-protobuf-transaction-encoding.md#unknown-field-filtering),
+sticking with the `Keeper` paradigm instead of ADR 033 may be unviable to begin with.
+
+A more performant solution (that could maybe be adapted to work with `Keeper` interfaces) would be to only expose
+getters and setters for generated types and internally store data in memory buffers which could be passed from
+one implementation to another in a zero-copy way.
+
+For example, imagine this protobuf API with only getters and setters is exposed for `MsgSend`:
+```go
+type MsgSend interface {
+	proto.Message
+	GetFromAddress() string
+	GetToAddress() string
+	GetAmount() []v1beta1.Coin
+    SetFromAddress(string)
+    SetToAddress(string)
+    SetAmount([]v1beta1.Coin)
+}
+
+func NewMsgSend() MsgSend { return &msgSendImpl{memoryBuffers: ...} }
+```
+
+Under the hood, `MsgSend` could be implemented based on some raw memory buffers, ex:
+```go
+type msgSendImpl struct {
+	memoryBuffers []interface{}
+}
+
+func (m *msgSendImpl) GetFromAddress() {
+	return m.memoryBuffers[0].(string)
+}
+
+func (m *msgSendImpl) GetToAddress() {
+    return m.memoryBuffers[1].(string)
+}
```

Comment:
`[]interface{}` describes a slice of `interface{}`. Slices are represented in memory as a 3-tuple of a pointer to a backing array, a length value reflecting the number of active elements in that array, and a capacity representing the total contiguous size of the array. An `interface{}` describes an interface type, which is represented in memory as a 2-tuple of a pointer to the concrete value implementing the interface, and a pointer to type information describing the properties of that interface.

memoryBuffers[0] will resolve to a value of type `interface{}`, which is a 2-tuple of pointer and type metadata. If that interface value was explicitly assigned a value of type `string` then the type metadata part of the interface 2-tuple would say "string" and this `.(string)` type assertion would work just fine. But if memoryBuffers is "over-laid" C-style on top of some arbitrary sequence of bytes in memory, this is not the case. There is no guarantee that the memory that happens to be in the spot that should represent the type metadata is valid, or represents the string type. And that means you can't make blind type assertions like this, at a minimum you need to check them with `, ok`. 

This is all ultimately a subversion of the type system. Why? Performance? The SDK certainly has tons of performance issues worth improving, but optimizations like this, trying to avoid copying tiny values like a MsgSend, don't even begin to register on that list. Performance is not impacted at all by stuff like this. 

Please just write plain dumb idiomatic Go. If you want to make the SDK faster, use the profiler.

Reply:
What is in the ADR is just a dummy implementation of this. I'll remove it and simply say that if we did this it would follow pretty closely whatever CapnProto/FlatBuffers do which AFAIK can be done safely within a language and across a VM boundary. One of the motivations is to make performance better if people want modules living in a VM (like WASM) which for better or worse seems like it's where things are going. Seems like you've flagged a pretty significant performance overhead to CosmWasm invocations (if you have specific benchmarks I'd love to see actually). The idea here would be limiting the overhead due to serialization over the WASM/inter-module boundary

Reply:
Have you looked at profiles of Cosmos apps that frequently traverse the "VM" boundary between Go and CosmWasm?

I have. At length, and in excruciating and mind-numbing detail. Things may have changed since I last looked, but as of a month or two ago, serialization costs had statistically zero (0) impact on net performance.

Reply:
Interesting, are there any published benchmarks?

Reply:
1. Start up a full node for any Cosmos network that leverages CosmWasm e.g. Osmosis
2. Enable the Prometheus endpoints in whatever .toml controls that
3. `go tool pprof -http=: <binary> http://localhost:26660/debug/pprof/profile\?seconds=30`
4. View -> Flamegraph, Filter -> "wasm"
5. Sigh deeply and reconsider, yet again, the virtues of becoming a woodworker or cordwainer

Reply:
I'm also interested. @peterbourgon do you have this setup handy to (re)-run and post a picture of the flame graph?